crates/cli/src/commands/upgrade.rs

```diff
@@ -0,0 +1,82 @@
+use std::{env, io, os::unix::prelude::PermissionsExt};
+
+use crate::helpers::{create_progress_bar, AnyError};
+use futures_util::TryStreamExt;
+use moon_launchpad::check_version;
+use moon_logger::error;
+use tokio::{
+    fs::{create_dir_all, rename, File},
+    io::copy,
+};
+use tokio_util::compat::FuturesAsyncReadCompatExt;
+
+pub async fn upgrade() -> Result<(), AnyError> {
+    let version = env!("CARGO_PKG_VERSION");
+    let version_check = check_version(version).await;
+
+    let new_version = match version_check {
+        Ok(Some(newer_version)) => newer_version,
+        Err(err) => {
+            error!("Failed to get current version of the cli from remote: {err}");
+            return Err(err);
+        }
+        Ok(None) => {
+            println!("You're already running the latest version of moon: {version}");
+            return Ok(());
+        }
+    };
+
+    let target = match (std::env::consts::OS, std::env::consts::ARCH) {
+        ("linux", arch) => {
+            // Run ldd to check if we're running on musl
+            let output = std::process::Command::new("ldd")
+                .arg("--version")
+                .output()?;
+            let output = String::from_utf8(output.stdout)?;
+            let libc = match output.contains("musl") {
+                true => "musl",
+                false => "gnu",
+            };
+            format!("moon-{arch}-unknown-linux-{libc}")
+        }
+        ("macos", arch) => format!("moon-{arch}-apple-darwin"),
+        ("windows", "x86_64") => "moon-x86_64-pc-windows-msvc.exe".to_owned(),
+        _ => unreachable!(),
+    };
+
+    let done = create_progress_bar(format!("Upgrading moon to version {new_version}..."));
+
+    let bin_path = env::current_exe()?;
+
+    // Move the old binary to a versioned path
+    let ver_path = bin_path
+        .parent()
+        .unwrap()
+        .join(&version)
+        .join(bin_path.file_name().unwrap());
+    create_dir_all(ver_path.parent().unwrap()).await?;
```

Comment:
use the `moon_utils::fs` methods for most (if not all) fs operations, as it has built in io handling